cloud-mask/cloud_mask.py

```diff
@@ -0,0 +1,190 @@
+import cv2
+import matplotlib.pyplot as plt
+import numpy as np
+from numpy.typing import NDArray
+from typing import Any, cast, Tuple
+import warnings
+
+
+class CloudMask:
+    def __init__(
+        self,
+        bands: NDArray[np.uint8] = None,
+        tif_path: str = None,
+        downsample_factor: float = 0.1,
+        ndsi_threshold: float = 0.3,
+        brightness_threshold: float = 0.5,
+        thermal_threshold: float = 0.5,  # example threshold
+    ) -> None:
+        """
+        Initialise using 4-band NDArray (default) or .tif file.
+
+        Parameters:
+            bands: The four bands of the image, as a 4xMxN tensor. Having the
+                input as an array ensures M and N are consistent for each band.
+            nsdi_threshold: Any pixel with an nsdi value above this threshold
+                will be masked out.
+            brightness_threshold: Any pixel with a brightness value above this
+                threshold will be masked out.
+            thermal_threshold: Any pixel with a thermal value above this
+                threshold will be masked out.
+        """
+        self.__band_red: NDArray[np.uint8] = None  # R
+        self.__band_green: NDArray[np.uint8] = None  # G
+        self.__band_blue: NDArray[np.uint8] = None  # B
+        self.__band_nir: NDArray[np.uint8] = None  # NIR
+
+        self.downsample_factor = downsample_factor
+        self.ndsi_threshold = ndsi_threshold
+        self.brightness_threshold = brightness_threshold
+        self.thermal_threshold = thermal_threshold
+
+        if bands is not None:
+            self.__band_red = bands[0]
+            self.__band_green = bands[1]
+            self.__band_blue = bands[2]
+            self.__band_nir = bands[3]
+        elif tif_path is not None:
+            (
+                self.__band_red,
+                self.__band_green,
+                self.__band_blue,
+                self.__band_nir,
+            ) = self.__readfile(tif_path, self.downsample_factor)
+        else:
+            raise ValueError(
+                "Cannot initialise CloudMask: no bands data or tif file path specified."
+            )
+
+    def __readfile(
+        self, file: str, downsample_factor: float
+    ) -> Tuple[np.ndarray, np.ndarray, np.ndarray, np.ndarray]:
+        """Load image bands from BGR image file"""
+        image_bgr: NDArray[np.uint8] = cv2.imread(file)
+
+        if image_bgr is None:
+            raise FileNotFoundError(
+                f"Could not load file '{file}'. Check file exists and is of '.tif' format."
+            )
+
+        # Compressing file for speed
+        new_size: Tuple[int, int] = (
+            int(image_bgr.shape[0] * downsample_factor),
+            int(image_bgr.shape[1] * downsample_factor),
+        )
+
+        image_resized: NDArray[np.uint8] = cv2.resize(
```

Comment:
This will also throw mypy errors. `cv2.resize` has return type `ndarray[Any, dtype[integer[Any] | floating[Any]]]`. You can fix this by casting the return value using `typing.cast`, but before we do this we should ensure that `cv2.resize` will actually return the type we want in this case. Based on [the documentation](https://docs.opencv.org/4.x/da/d54/group__imgproc__transform.html#ga47a974309e9102f5f08231edc7e7529d), it will return an array of type `uint8` unless I misunderstand: "the type of dst is the same as of src." So the cast is okay to do.